python/triton/experimental/gluon/language/_semantic.py

```diff
@@ -267,6 +267,51 @@ def shared_load(self, mem_desc, layout):
         handle = self.builder.create_local_load(ret_ty.to_ir(self.builder), mem_desc.handle)
         return ttgl.tensor(handle, ret_ty)
 
+    def _store_legacy(self, ptr, val, mask, boundary_check, cache, eviction):
+        # Store by a tensor of pointers or a pointer of scalar: `block_type<pointer_type<>>` or `pointer_type<>`
+        if not ptr.type.scalar.is_ptr():
+            raise ValueError(f"Unsupported ptr type {ptr.type.__repr__()} in `tl.store`")
+
+        # Check `boundary_check` argument
+        if boundary_check:
+            raise ValueError("`boundary_check` argument is not supported for storing a tensor of pointers or storing a "
+                             "scalar. Because the compiler does not know the boundary; please use block pointers "
+                             "(defined by `make_block_ptr`) instead")
+
+        # For a pointer of scalar, check the type of `val` and `mask`
+        if not ptr.type.is_block():
+            if val.type.is_block():
+                raise ValueError("Value argument cannot be block type if pointer argument is not a block")
+            if mask and mask.type.is_block():
+                raise ValueError("Mask argument cannot be block type if pointer argument is not a block")
+
+        # Make `mask` and `val` into the same shape as `ptr`
+        if ptr.type.is_block():
+            # TODO fix python/test/gluon/test_lowerings.py
+            ptr, val = self.broadcast_impl_value(ptr, val)
+            if mask is not None:
+                ptr, mask = self.broadcast_impl_value(ptr, mask)
```

Comment:
Please don't duplicate this code, just update the triton semantic to use `broadcast_impl_value` and it will work for both. Also, there is still a bug here if `ptr` and `val` have `AutoLayout` but `mask` has a concrete layout. The fix is to call `self.broadcast_tensors(ptr, val, mask)`

Reply:
What are the test failures? I would expect that if it's broken for triton it would also be broken for gluon.